topic/machine-learning/llm-langchain/cratedb_rag_customer_support_langchain.ipynb

```diff
@@ -410,11 +460,160 @@
+      ]
+     },
+     "execution_count": 48,
+     "metadata": {},
+     "output_type": "execute_result"
+    }
+   ],
    "source": [
     "chat_completion.choices[0].message.content"
    ]
   },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## CrateDB &#129505; OpenSource\n",
+    "\n",
+    "In the second part of this notebook we build a Retrieval-Augmented Generation (RAG) system using CrateDB with Jina AI and the Llama open-source models. \n",
+    "\n",
+    "### Configure Jina AI API "
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "Jina AI embeddings on Hugging Face are a suite of open-source models designed for encoding text data into high-dimensional vectors, which can be used in various natural language processing tasks such as semantic search, text classification, and clustering. These embeddings utilize the BERT architecture and are specially tuned for longer sequence lengths up to 8192 tokens, which is notably higher than the standard BERT model's capacity.\n",
+    "\n",
+    "To use Jina AI embeddings you need a Jina AI API key which can be easily obtained once you sign up for an account on the Jina AI website."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 42,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "from langchain_community.embeddings import JinaEmbeddings\n",
+    "\n",
+    "getenvpass(\"JINA_API_KEY\", prompt=\"Jina API key:\")\n",
+    "embeddings = JinaEmbeddings()"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "### Initialize new embedding store\n",
+    "\n",
+    "The next step initializes a vector search store in CrateDB using embeddings generated by Jina AI models and returns relevant documents using similarity search."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 50,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "store = CrateDBVectorSearch.from_documents(\n",
+    "    embedding=embeddings,\n",
+    "    documents=data,\n",
+    "    collection_name=COLLECTION_NAME,\n",
+    "    connection_string=CONNECTION_STRING,\n",
+    ")\n",
+    "documents = return_documents(store, my_question)"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "### Connect to a local LLM "
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "One of the easiest ways to connect to a local LLM is to download one of the models from the Llamafile repository (https://github.com/Mozilla-Ocho/llamafile?tab=readme-ov-file) and run it locally. Upon initiating the llamafile, it not only sets up a web-based user interface for chatting at the local address http://127.0.0.1:8080/, but it also offers an endpoint for chat completions that is compatible with the OpenAI API.\n",
+    "\n",
+    "In this example, we use the Phi-2 model that is about 2GB in size."
```

Comment:
> connect to a local LLM is to download one of the models

It looks like this may need additional infrastructure. Do you think it makes sense to demonstrate that whole topic on behalf of a separate / dedicated notebook? We are only just asking: If you think it is absolutely needed for semantic purposes to have it inside here, we will obey. Otherwise, we are asking to reconsider.

> One of the easiest ways to connect to a local LLM is to download one of the models from the llamafile repository.

What about using [Ollama](https://github.com/ollama/ollama)? Is it a viable alternative to that procedure using [llamafile](https://github.com/Mozilla-Ocho/llamafile)?

Reply:
> What about using [....] to acquire and run an LLM model locally?

Oh wow. That description about llamafile sounds intriguing, indeed. 🍀 

### About llamafile

> llamafile lets you distribute and run LLMs with a single file. ([announcement blog post](https://hacks.mozilla.org/2023/11/introducing-llamafile/))

> Our goal is to make open LLMs much more accessible to both developers and end users. We're doing that by combining [llama.cpp](https://github.com/ggerganov/llama.cpp) with [Cosmopolitan Libc](https://github.com/jart/cosmopolitan) into one framework that collapses all the complexity of LLMs down to a single-file executable (called a "llamafile") that runs locally on most computers, with no installation.

Reply:
There is a overlap in supported models when it comes to Ollama. For instance, it is very convenient to run Phi-2 locally and the Mozilla repository gives a nice set of executable files that have API endpoint compatible with the OpenAI interface. 

For the upcoming presentation, I insist having both examples in a single file. After that we can create a new 'open source' example and play with other models too.

Reply:
> For the upcoming presentation, I insist having both examples in a single file. After that we can [refactor] ...

Acknowledged. Thanks for clarifying.